File: Data_Structures/stupid_heap.cpp
#include <algorithm>
#include <climits>

constexpr int MAXN = 100;

// The simplest & slowest & dumbest heap implementation
struct heap_v1 {
private:
    int h[MAXN];
    int n = 0;  // current size of the heap

public:
    // insert an elemenet at the end. O(1)
    void insert(int x) {
        h[n++] = x;
    }

    // get minimum element in the heap. O(n)
    int get_min() const {
        int res = INT_MAX;
        for (int i = 0; i < n; i++) {
            res = std::min(res, h[i]);
        }
        return res;
    }

    // remove min of the heap. O(n)
    void remove_min() {
        int j = 0;
        for (int i = 1; i < n; i++) {
            if (h[j] > h[i]) j = i;
        }
        std::swap(h[j], h[n-1]);
        n--;
    }
};

// heap with storing the min element
struct heap_v2 {
private:
    int h[MAXN];
    int n = 0;
    int m;

public:
    // O(1)
    int get_min() const {
        return h[m];
    }

    // O(n)
    void remove_min() {
        std::swap(h[m], h[n-1]);
        n--;
        // update new minimum
        m = 0;
        for (int i = 1; i < n; i++) {
            if (h[i] < h[m]) m = i;
        }
    }

    // O(1)
    void insert(int x) {
        h[n++] = x;
        if (x < h[m]) m = n-1;
    }
};

// heap with fast minimum element deletion
struct heap_v3 {
private:
    int h[MAXN];
    int n = 0;

public:
    // O(1)
    void remove_min() {
        n--;    // careful no out of bounds check here
    }

    // O(1)
    int get_min() {
        return h[n--];
    }

    // sift-left (1-run of insertion sort). O(n)
    void insert(int x) {
        h[n++] = x;
        int j = n-1;
        while (j > 0 && h[j] > h[j-1]) {
            std::swap(h[j], h[j-1]);
            j--;
        }
    }
};

File: datastructures/gol/heap/stupid_heap.go
package heap

import "math"

// The simplest & slowest & dumbest heap implementation
type HeapV1 struct {
	h []int
}

// insert an elemenet at the end. O(1)
func (h *HeapV1) Insert(x int) {
	h.h = append(h.h, x)
}

// get minimum element in the heap. O(n)
func (h *HeapV1) GetMin() (int, bool) {
	if len(h.h) == 0 {
		return 0, false
	}
	res := math.MaxInt
	for _, x := range h.h {
		if x < res {
			res = x
		}
	}
	return res, true
}

// remove min of the heap. O(n)
func (h *HeapV1) RemoveMin() bool {
	if len(h.h) == 0 {
		return false
	}
	j := 0
	for i := 1; i < len(h.h); i++ {
		if h.h[j] > h.h[i] {
			j = i
		}
	}
	last := len(h.h) - 1
	h.h[j], h.h[last] = h.h[last], h.h[j]
	h.h = h.h[:last]
	return true
}

// heap with storing the min element
type HeapV2 struct {
	h []int
	m int
}

// O(1)
func (h *HeapV2) GetMin() (int, bool) {
	if len(h.h) == 0 {
		return 0, false
	}
	return h.h[h.m], true
}

// O(n)
func (h *HeapV2) RemoveMin() bool {
	if len(h.h) == 0 {
		return false
	}
	last := len(h.h) - 1
	h.h[h.m], h.h[last] = h.h[last], h.h[h.m]
	h.h = h.h[:last]

	// update new minimum
	h.m = 0
	for i := 1; i < len(h.h); i++ {
		if h.h[i] < h.h[h.m] {
			h.m = i
		}
	}
	return true
}

// O(1)
func (h *HeapV2) Insert(x int) {
	h.h = append(h.h, x)
	if len(h.h) == 1 || x < h.h[h.m] {
		h.m = len(h.h) - 1
	}
}

// heap with fast minimum element deletion
type HeapV3 struct {
	h []int
}

// O(1)
func (h *HeapV3) RemoveMin() bool {
	if len(h.h) == 0 {
		return false
	}
	h.h = h.h[:len(h.h)-1] // careful no out of bounds check here in the original
	return true
}

// O(1)
func (h *HeapV3) GetMin() (int, bool) {
	if len(h.h) == 0 {
		return 0, false
	}
	return h.h[len(h.h)-1], true
}

// sift-left (1-run of insertion sort). O(n)
func (h *HeapV3) Insert(x int) {
	h.h = append(h.h, x)
	j := len(h.h) - 1
	for j > 0 && h.h[j] > h.h[j-1] {
		h.h[j], h.h[j-1] = h.h[j-1], h.h[j]
		j--
	}
}
